src/main/scala/workflow/Pipeline.scala

```diff
@@ -0,0 +1,197 @@
+    } :+ s"$sink -> ${nodes.size}"
+
+    val fitEdges: Seq[String] = fitDeps.zipWithIndex.flatMap {
+      case (deps, id) => deps.map(x => s"$x -> $id [dir=${"\"none\""} style=${"\"dashed\""}]")
+    }
+
+    val ranks = fitDeps.zipWithIndex.flatMap {
+      case (deps, id) => deps.map(x => s"{rank=same; $x $id}")
+    }
+
+    val lines = nodeLabels ++ dataEdges ++ fitEdges ++ ranks
+    lines.mkString("digraph pipeline {\n  rankdir=LR;\n  ", "\n  ", "\n}")
+  }
+
+  final private[workflow] def planEquals(pipeline: Pipeline[A, B]): Boolean = {
+    this.eq(pipeline) || (
+        (nodes == pipeline.nodes) &&
+        (dataDeps == pipeline.dataDeps) &&
+        (fitDeps == pipeline.fitDeps) &&
+        (sink == pipeline.sink))
+  }
+}
+
+object Pipeline {
+  val SOURCE: Int = -1
+
+  def apply[T](): Pipeline[T, T] = new ConcretePipeline(Seq(), Seq(), Seq(), SOURCE)
+  private[workflow] def apply[A, B](nodes: Seq[Node], dataDeps: Seq[Seq[Int]], fitDeps: Seq[Seq[Int]], sink: Int): Pipeline[A, B] = new ConcretePipeline(nodes, dataDeps, fitDeps, sink)
+}
+
+private[workflow] class ConcretePipeline[A, B](
+  private[workflow] override val nodes: Seq[Node],
+  private[workflow] override val dataDeps: Seq[Seq[Int]],
+  private[workflow] override val fitDeps: Seq[Seq[Int]],
+  private[workflow] override val sink: Int) extends Pipeline[A, B] with Logging {
+
+  private val fitCache: Array[Option[TransformerNode[_]]] = nodes.map(_ => None).toArray
+
+  private def fitEstimator(node: Int): TransformerNode[_] = fitCache(node).getOrElse {
+    nodes(node) match {
+      case _: DataNode =>
+        throw new RuntimeException("Pipeline DAG error: Cannot have a fit dependency on a DataNode")
+      case _: TransformerNode[_] =>
+        throw new RuntimeException("Pipeline DAG error: Cannot have a data dependency on a Transformer")
+      case estimator: EstimatorNode =>
+        val nodeDataDeps = dataDeps(node).map(x => rddDataEval(x, null))
+        logInfo(s"Fitting '${estimator.label}' [$node]")
+        val fitOut = estimator.fit(nodeDataDeps)
+        fitCache(node) = Some(fitOut)
+        logInfo(s"Finished fitting '${estimator.label}' [$node]")
+        fitOut
+    }
+  }
+
+  private def singleDataEval(node: Int, in: A): Any = {
+    if (node == Pipeline.SOURCE) {
+      in
+    } else {
+      nodes(node) match {
+        case transformer: TransformerNode[_] =>
+          val nodeFitDeps = fitDeps(node).map(fitEstimator)
+          val nodeDataDeps = dataDeps(node).map(x => singleDataEval(x, in))
+          transformer.transform(nodeDataDeps, nodeFitDeps)
+        case _: DataNode =>
+          throw new RuntimeException("Pipeline DAG error: came across an RDD data dependency when trying to do a single item apply")
+        case _: EstimatorNode =>
+          throw new RuntimeException("Pipeline DAG error: Cannot have a data dependency on an Estimator")
+      }
+    }
+  }
+
+  private def rddDataEval(node: Int, in: RDD[A]): RDD[_] = {
```

Comment:
I'm planning to use this (and possibly its non-RDD brother) in my caching stuff. Any good reason not to mark it private[workflow]?